avr-hal-generic/src/spi.rs

```diff
@@ -266,6 +306,59 @@ where
     }
 }
 
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN> embedded_hal::spi::ErrorType
+    for Spi<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+    CSPIN: port::PinOps,
+{
+    type Error = core::convert::Infallible;
+}
+
+impl<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN> embedded_hal::spi::SpiBus
+    for Spi<H, SPI, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>
+where
+    SPI: SpiOps<H, SCLKPIN, MOSIPIN, MISOPIN, CSPIN>,
+    SCLKPIN: port::PinOps,
+    MOSIPIN: port::PinOps,
+    MISOPIN: port::PinOps,
+    CSPIN: port::PinOps,
+{
+    fn read(&mut self, read: &mut [u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_read_buf(read))
+    }
+
+    fn write(&mut self, write: &[u8]) -> Result<(), Self::Error> {
+        Ok(self.p.raw_write_buf(write))
+    }
+
+    fn transfer(&mut self, read: &mut [u8], write: &[u8]) -> Result<(), Self::Error> {
+        let common_length = read.len().min(write.len()); 
+        let (write1, write2) = write.split_at(common_length); 
+        let (read1, read2) = read.split_at_mut(common_length); 
+        self.p.raw_read_write_buf( read1, write1); 
+        
+        debug_assert!(write2.len() == 0 || read2.len() == 0); 
+        self.p.raw_write_buf(write2); 
+        self.p.raw_read_buf(read2);
+        Ok(())
+    }
+    fn transfer_in_place(&mut self, buffer: &mut [u8]) -> Result<(), Self::Error> {
+        self.p.raw_read_write_in_place(buffer);
+        Ok(())
+    }
+    fn flush(&mut self) -> Result<(), Self::Error> {
+        if self.write_in_progress {
+            while !self.p.raw_check_iflag() {}
+            self.write_in_progress = false;
+        }
+        Ok(())
```

Comment:
This is the only thing I'm still not entirely sure about:  As outlined in comment https://github.com/Rahix/avr-hal/pull/503#discussion_r1485565042, I see potential issues in the interaction of the new methods with the old ones.  While the new methods always leave the peripheral in an idle state, the old ones do not.  So we need to prepare for situations where both kinds are called from a single program.

I think the most simple solution for now is to just call `self.flush()` at the beginning of all the new methods (`SpiBus::{read,write,transfer,transfer_in_place}`).  With a comment to explain why, for example:

```rust
    fn write(&mut self, write: &[u8]) -> Result<(), Self::Error> {
        // Must flush() first, if asynchronous operations happened before this
        self.flush().unwrap_infallible();
        Ok(self.p.raw_write_buf(write))
    }
```